File: solutions/R/ch18_ex3.R
# =============================================================================
# Chapter 18 - Exercise 3: Subgroup Analysis and Meta-Regression in R
# =============================================================================

library(tidyverse)
library(meta)
library(metafor)

# --- Dataset ---
af_trials <- data.frame(
  study = c("TRAIL-1", "GUARD-AF", "SHIELD", "ORBIT-AF",
            "VENTURE", "COMPASS-AF", "PIONEER-2", "ATLAS-AF"),
  events_new  = c(28, 45, 112, 67, 33, 89, 52, 41),
  n_new       = c(1200, 2500, 5400, 3100, 1800, 4200, 2800, 2100),
  events_warf = c(42, 58, 148, 84, 29, 102, 61, 53),
  n_warf      = c(1200, 2500, 5400, 3100, 1800, 4200, 2800, 2100),
  region      = c("Europe", "North America", "Europe", "Asia",
                   "North America", "Europe", "Asia", "North America"),
  mean_age    = c(72, 68, 74, 65, 70, 71, 63, 69),
  pct_female  = c(38, 42, 35, 48, 40, 37, 52, 44)
)

# Compute effect sizes using metafor
es <- escalc(measure = "RR",
             ai = events_new, n1i = n_new,
             ci = events_warf, n2i = n_warf,
             data = af_trials)

# Base meta-analysis
m1 <- metabin(
  event.e = events_new, n.e = n_new,
  event.c = events_warf, n.c = n_warf,
  studlab = study, data = af_trials,
  sm = "RR", method.tau = "REML"
)

# --- Part (a): Subgroup analysis by region ---
cat("--- Part (a): Subgroup analysis by region ---\n\n")

# Using meta package
m_sub <- update(m1, subgroup = af_trials$region)
cat("Subgroup analysis results:\n")
print(summary(m_sub))

# Forest plot by subgroup
forest(m_sub,
       sortvar = TE,
       label.left = "Favours new anticoagulant",
       label.right = "Favours warfarin",
       col.diamond = "steelblue",
       print.tau2 = TRUE,
       print.I2 = TRUE,
       print.subgroup.name = TRUE,
       main = "Subgroup Analysis by Region")

# Using metafor for the test of subgroup differences
res_sub <- rma(yi, vi, mods = ~ region, data = es, method = "REML")
cat("\nTest for subgroup differences (metafor):\n")
print(summary(res_sub))

cat("\nDo treatment effects differ by region?\n")
cat("Test for moderation: QM =", round(res_sub$QM, 2),
    ", df =", res_sub$m, ", p =", round(res_sub$QMp, 4), "\n")
if (res_sub$QMp < 0.05) {
  cat("Yes, there is a statistically significant difference between regions.\n")
} else {
  cat("No, there is no statistically significant difference between regions.\n")
  cat("However, with only", nrow(es), "studies split across",
      length(unique(es$region)), "regions,\n")
  cat("statistical power for detecting subgroup differences is limited.\n")
}

# --- Part (b): Meta-regression with mean age ---
cat("\n--- Part (b): Meta-regression with mean age ---\n\n")

res_age <- rma(yi, vi, mods = ~ mean_age, data = es, method = "REML")
print(summary(res_age))

cat("\nInterpretation:\n")
cat("Coefficient for mean_age:", round(coef(res_age)["mean_age"], 4), "\n")
cat("p-value:", round(res_age$pval[2], 4), "\n")
cat("R^2 (proportion of heterogeneity explained):",
    round(max(0, res_age$R2), 1), "%\n")

if (res_age$pval[2] < 0.05) {
  cat("There IS a statistically significant relationship between mean age\n")
  cat("and treatment effect. Each 1-year increase in mean age is associated\n")
  cat("with a", round(coef(res_age)["mean_age"], 4),
      "change in log RR.\n")
} else {
  cat("There is NO statistically significant relationship between mean age\n")
  cat("and treatment effect (p =", round(res_age$pval[2], 3), ").\n")
}

# --- Part (c): Meta-regression with percentage female ---
cat("\n--- Part (c): Meta-regression with percentage female ---\n\n")

res_fem <- rma(yi, vi, mods = ~ pct_female, data = es, method = "REML")
print(summary(res_fem))

cat("\nInterpretation:\n")
cat("Coefficient for pct_female:", round(coef(res_fem)["pct_female"], 4), "\n")
cat("p-value:", round(res_fem$pval[2], 4), "\n")
cat("R^2:", round(max(0, res_fem$R2), 1), "%\n")

cat("\n*** ECOLOGICAL FALLACY WARNING ***\n")
cat("Even if there is an association between trial-level percentage female\n")
cat("and the treatment effect, this does NOT prove that individual women\n")
cat("respond differently to the treatment than individual men.\n")
cat("Trials with higher percentage female may differ in other ways:\n")
cat("  - Geographic region (cultural differences in enrolment)\n")
cat("  - Mean age (women often present with AF at older ages)\n")
cat("  - Comorbidity profiles\n")
cat("  - Trial methodology\n")
cat("Only an individual participant data (IPD) meta-analysis with a\n")
cat("treatment-by-sex interaction term can properly assess whether\n")
cat("sex modifies the treatment effect.\n")

# --- Part (d): Bubble plot of meta-regression on mean age ---
cat("\n--- Part (d): Bubble plot ---\n")

# Method 1: Using metafor's built-in bubble plot
regplot(res_age,
        xlab = "Mean Age (years)",
        ylab = "Log Risk Ratio",
        main = "Meta-Regression: Treatment Effect vs Mean Age",
        ci = TRUE,
        pi = TRUE,  # prediction interval
        col = "steelblue",
        bg = "lightblue",
        las = 1)

# Method 2: Using ggplot2 for more control
es$w_re <- 1 / (es$vi + res_age$tau2)
es$w_norm <- es$w_re / max(es$w_re)

# Predicted line from meta-regression
age_seq <- seq(min(es$mean_age) - 2, max(es$mean_age) + 2, length.out = 100)
pred <- predict(res_age, newmods = age_seq)

pred_df <- data.frame(
  mean_age = age_seq,
  pred = pred$pred,
  ci_lo = pred$ci.lb,
  ci_hi = pred$ci.ub,
  pi_lo = pred$pi.lb,
  pi_hi = pred$pi.ub
)

p_bubble <- ggplot() +
  # Prediction interval
  geom_ribbon(data = pred_df, aes(x = mean_age, ymin = pi_lo, ymax = pi_hi),
              fill = "grey90", alpha = 0.5) +
  # Confidence interval for regression line
  geom_ribbon(data = pred_df, aes(x = mean_age, ymin = ci_lo, ymax = ci_hi),
              fill = "steelblue", alpha = 0.2) +
  # Regression line
  geom_line(data = pred_df, aes(x = mean_age, y = pred),
            colour = "steelblue", linewidth = 1) +
  # Study points (bubbles)
  geom_point(data = es, aes(x = mean_age, y = yi, size = w_norm),
             colour = "steelblue", alpha = 0.7) +
  # Study labels
  geom_text(data = es, aes(x = mean_age, y = yi, label = study),
            size = 2.5, vjust = -1.5) +
  # Reference line at null
  geom_hline(yintercept = 0, linetype = "dashed", colour = "grey50") +
  # Aesthetics
  scale_size_continuous(range = c(3, 10), guide = "none") +
  labs(x = "Mean Age (years)",
       y = "Log Risk Ratio",
       title = "Bubble Plot: Meta-Regression of Treatment Effect on Mean Age",
       subtitle = paste0("Slope = ", round(coef(res_age)["mean_age"], 4),
                        ", p = ", round(res_age$pval[2], 3),
                        "; bubble size proportional to study weight")) +
  theme_minimal() +
  theme(
    plot.title = element_text(face = "bold", size = 12),
    axis.title = element_text(face = "bold")
  )

print(p_bubble)

ggsave("bubble_plot_age.png", p_bubble, width = 8, height = 6, dpi = 300)

cat("\nBubble plot saved. Larger bubbles indicate more precise studies.\n")
cat("The regression line shows the relationship between mean age\n")
cat("and treatment effect, with shaded confidence and prediction intervals.\n")

File: solutions/python/ch18_ex3.py
# =============================================================================
# Chapter 18 - Exercise 3: Subgroup Analysis and Meta-Regression in Python
# =============================================================================

import numpy as np
import pandas as pd
from scipy import stats
import statsmodels.api as sm
import matplotlib.pyplot as plt

# --- Dataset ---
af_trials = pd.DataFrame({
    'study': ['TRAIL-1', 'GUARD-AF', 'SHIELD', 'ORBIT-AF',
              'VENTURE', 'COMPASS-AF', 'PIONEER-2', 'ATLAS-AF'],
    'events_new':  [28, 45, 112, 67, 33, 89, 52, 41],
    'n_new':       [1200, 2500, 5400, 3100, 1800, 4200, 2800, 2100],
    'events_warf': [42, 58, 148, 84, 29, 102, 61, 53],
    'n_warf':      [1200, 2500, 5400, 3100, 1800, 4200, 2800, 2100],
    'region':      ['Europe', 'North America', 'Europe', 'Asia',
                    'North America', 'Europe', 'Asia', 'North America'],
    'mean_age':    [72, 68, 74, 65, 70, 71, 63, 69],
    'pct_female':  [38, 42, 35, 48, 40, 37, 52, 44]
})

# Compute effect sizes
af_trials['rr'] = (af_trials['events_new'] / af_trials['n_new']) / \
                  (af_trials['events_warf'] / af_trials['n_warf'])
af_trials['log_rr'] = np.log(af_trials['rr'])
af_trials['var_log_rr'] = (
    1/af_trials['events_new'] - 1/af_trials['n_new'] +
    1/af_trials['events_warf'] - 1/af_trials['n_warf']
)
af_trials['se_log_rr'] = np.sqrt(af_trials['var_log_rr'])


def dl_random_effects(log_rr, var):
    """DerSimonian-Laird random-effects meta-analysis."""
    k = len(log_rr)
    w_fe = 1 / var
    pooled_fe = np.sum(w_fe * log_rr) / np.sum(w_fe)
    Q = np.sum(w_fe * (log_rr - pooled_fe)**2)
    C = np.sum(w_fe) - np.sum(w_fe**2) / np.sum(w_fe)
    tau2 = max(0, (Q - (k - 1)) / C)
    w_re = 1 / (var + tau2)
    pooled_re = np.sum(w_re * log_rr) / np.sum(w_re)
    se_re = np.sqrt(1 / np.sum(w_re))
    I2 = max(0, (Q - (k-1)) / Q) * 100 if Q > 0 else 0
    return pooled_re, se_re, tau2, Q, I2, w_re


# Overall random-effects analysis
pooled_re, se_re, tau2, Q, I2, w_re = dl_random_effects(
    af_trials['log_rr'].values, af_trials['var_log_rr'].values
)

# --- Part (a): Subgroup analysis by region ---
print("=" * 60)
print("Part (a): Subgroup Analysis by Region")
print("=" * 60)

regions = af_trials['region'].unique()
subgroup_results = []

for region in regions:
    mask = af_trials['region'] == region
    sub = af_trials[mask]

    if len(sub) >= 2:
        p_re, s_re, t2, q, i2, w = dl_random_effects(
            sub['log_rr'].values, sub['var_log_rr'].values
        )
        ci = (p_re - 1.96*s_re, p_re + 1.96*s_re)
    else:
        # Single study: use study estimate directly
        p_re = sub['log_rr'].values[0]
        s_re = sub['se_log_rr'].values[0]
        t2, q, i2 = 0, 0, 0
        ci = (p_re - 1.96*s_re, p_re + 1.96*s_re)

    subgroup_results.append({
        'region': region,
        'k': len(sub),
        'pooled_rr': np.exp(p_re),
        'ci_lo': np.exp(ci[0]),
        'ci_hi': np.exp(ci[1]),
        'I2': i2,
        'tau2': t2
    })

    print(f"\n{region} (k={len(sub)}):")
    print(f"  Pooled RR: {np.exp(p_re):.4f} "
          f"(95% CI: {np.exp(ci[0]):.4f} - {np.exp(ci[1]):.4f})")
    print(f"  I^2: {i2:.1f}%")

# Test for subgroup differences using meta-regression with region dummies
region_dummies = pd.get_dummies(af_trials['region'], drop_first=True).astype(float)
X_sub = sm.add_constant(region_dummies)
wls_sub = sm.WLS(af_trials['log_rr'], X_sub,
                 weights=1/af_trials['var_log_rr']).fit()

# Wald test for region coefficients
from scipy.stats import chi2
f_stat = wls_sub.f_test(np.eye(len(wls_sub.params))[1:])
print(f"\nTest for subgroup differences:")
print(f"  F-statistic: {f_stat.fvalue[0][0]:.3f}")
print(f"  p-value: {f_stat.pvalue:.4f}")

if f_stat.pvalue < 0.05:
    print("  Significant difference between regions.")
else:
    print("  No significant difference between regions.")
    print(f"  (Limited power with only {len(af_trials)} studies)")

# --- Part (b): Meta-regression with mean age ---
print(f"\n{'=' * 60}")
print("Part (b): Meta-Regression with Mean Age")
print("=" * 60)

X_age = sm.add_constant(af_trials['mean_age'])
wls_age = sm.WLS(af_trials['log_rr'], X_age,
                 weights=1/af_trials['var_log_rr']).fit()

print(f"\nMeta-regression: log_rr ~ mean_age")
print(f"  Intercept: {wls_age.params.iloc[0]:.4f} (p = {wls_age.pvalues.iloc[0]:.4f})")
print(f"  Mean age slope: {wls_age.params.iloc[1]:.4f} (p = {wls_age.pvalues.iloc[1]:.4f})")
print(f"  R-squared: {wls_age.rsquared:.3f}")

if wls_age.pvalues.iloc[1] < 0.05:
    print("\n  There IS a significant relationship between mean age and effect size.")
    if wls_age.params.iloc[1] < 0:
        print("  Trials with older populations show larger treatment benefits.")
    else:
        print("  Trials with older populations show smaller treatment benefits.")
else:
    print(f"\n  No significant relationship (p = {wls_age.pvalues.iloc[1]:.3f}).")

# --- Part (c): Meta-regression with percentage female ---
print(f"\n{'=' * 60}")
print("Part (c): Meta-Regression with Percentage Female")
print("=" * 60)

X_fem = sm.add_constant(af_trials['pct_female'])
wls_fem = sm.WLS(af_trials['log_rr'], X_fem,
                 weights=1/af_trials['var_log_rr']).fit()

print(f"\nMeta-regression: log_rr ~ pct_female")
print(f"  Intercept: {wls_fem.params.iloc[0]:.4f} (p = {wls_fem.pvalues.iloc[0]:.4f})")
print(f"  Pct female slope: {wls_fem.params.iloc[1]:.4f} (p = {wls_fem.pvalues.iloc[1]:.4f})")
print(f"  R-squared: {wls_fem.rsquared:.3f}")

print("\n  *** ECOLOGICAL FALLACY WARNING ***")
print("  This analysis examines TRIAL-LEVEL associations between percentage")
print("  female and treatment effect. Even if significant, this does NOT")
print("  prove that individual women respond differently to treatment.")
print("  Trials with more women may differ in other characteristics:")
print("    - Geographic region and healthcare systems")
print("    - Mean age and comorbidity profiles")
print("    - Trial methodology and inclusion criteria")
print("  Only individual participant data (IPD) meta-analysis with a")
print("  treatment-by-sex interaction can properly test effect modification.")

# --- Part (d): Bubble plot ---
print(f"\n{'=' * 60}")
print("Part (d): Bubble Plot")
print("=" * 60)

fig, ax = plt.subplots(figsize=(10, 7))

# Study weights for bubble sizes
w_plot = 1 / af_trials['var_log_rr']
sizes = (w_plot / w_plot.max()) * 500

# Scatter: bubbles
ax.scatter(af_trials['mean_age'], af_trials['log_rr'],
           s=sizes, alpha=0.6, c='steelblue', edgecolors='darkblue',
           linewidth=1, zorder=5)

# Add study labels
for _, row in af_trials.iterrows():
    ax.annotate(row['study'],
                (row['mean_age'], row['log_rr']),
                textcoords="offset points", xytext=(8, 8),
                fontsize=8, color='grey30')

# Meta-regression line
age_range = np.linspace(af_trials['mean_age'].min() - 2,
                        af_trials['mean_age'].max() + 2, 100)
pred = wls_age.params.iloc[0] + wls_age.params.iloc[1] * age_range
ax.plot(age_range, pred, 'steelblue', linewidth=2,
        label=f'Meta-regression line (slope={wls_age.params.iloc[1]:.4f})')

# Confidence band (approximate)
X_pred = sm.add_constant(age_range)
pred_full = wls_age.get_prediction(X_pred)
pred_ci = pred_full.conf_int(alpha=0.05)
ax.fill_between(age_range, pred_ci[:, 0], pred_ci[:, 1],
                alpha=0.15, color='steelblue', label='95% CI')

# Reference line
ax.axhline(y=0, color='grey', linestyle='--', linewidth=0.8,
           label='Null (log RR = 0)')

ax.set_xlabel('Mean Age (years)', fontweight='bold', fontsize=12)
ax.set_ylabel('Log Risk Ratio', fontweight='bold', fontsize=12)
ax.set_title('Bubble Plot: Meta-Regression of Treatment Effect on Mean Age\n'
             f'(Slope p = {wls_age.pvalues.iloc[1]:.3f}; '
             f'bubble size = study weight)',
             fontweight='bold', fontsize=13)
ax.legend(loc='best', frameon=True, framealpha=0.9)

plt.tight_layout()
plt.savefig('bubble_plot_age.png', dpi=300)
plt.show()

print("\nBubble plot saved as 'bubble_plot_age.png'")
print("Larger bubbles represent more precise (higher-weight) studies.")
